tools/test-stats/Main.hs

```diff
@@ -0,0 +1,171 @@
+failureCase = Report 0 1
+
+-- * XML parser
+
+-- | Returns (report by test case name, overallTestSuiteFailures, overallTestSuiteSucesses)
+--
+-- The overall numbers are always 1 or 0 and wrapped in the `Sum` type so we can
+-- mconcat them to get the overall sum.
+parse :: FilePath -> IO (MonoidalMap TestCase Report, Sum Int, Sum Int)
+parse fn = do
+  testCaseReport <- parseRaw <$> T.readFile fn
+  if any (\r -> r.failure > 0) testCaseReport
+    then pure (testCaseReport, Sum 1, Sum 0)
+    else pure (testCaseReport, Sum 0, Sum 1)
+
+parseRaw :: Text -> MonoidalMap TestCase Report
+parseRaw = mconcat . map parseTestSuites . selectTestSuites . parseXML
+  where
+    selectTestSuites = filter (matchQName "testsuites" . elName) . onlyElems
+
+parseTestSuites :: Element -> MonoidalMap TestCase Report
+parseTestSuites = mconcat . liftM2 map (parseTestSuite . name) (children "testsuite")
+
+parseTestSuite :: [Text] -> Element -> MonoidalMap TestCase Report
+parseTestSuite names =
+  liftM2
+    (MonoidalMap.unionWith (<>))
+    (mconcat . liftM2 map (parseTestSuite . (names <>) . name) (children "testsuite"))
+    (mconcat . liftM2 map (parseTestCase . (names <>) . name) (children "testcase"))
+
+parseTestCase :: [Text] -> Element -> MonoidalMap TestCase Report
+parseTestCase names el
+  | null failures = MonoidalMap.singleton compName successCase
+  | otherwise = MonoidalMap.singleton compName $ failureCase failures
+  where
+    compName = T.intercalate "." $ names <> name el
+    failures = map (T.pack . strContent) $ children "failure" el
+
+matchQName :: String -> QName -> Bool
+matchQName = flip $ (==) . qName
+
+children :: String -> Element -> [Element]
+children = filterChildrenName . matchQName
+
+name :: Element -> [Text]
+name = maybeToList . fmap T.pack . findAttrBy (matchQName "name")
+
+-- * Postgresql outout
+
+runMigrations :: Connection -> IO ()
+runMigrations conn = do
+  void $ execute_ conn "CREATE TABLE IF NOT EXISTS test_suite_runs (id SERIAL PRIMARY KEY, run_id VARCHAR, time TIMESTAMP, version VARCHAR, suite VARCHAR, failed_runs INT, successful_runs INT)"
+  void $ execute_ conn "CREATE TABLE IF NOT EXISTS test_case_runs (id SERIAL PRIMARY KEY, name VARCHAR, test_suite_run_id integer, failure_count integer, success_count integer)"
+  void $ execute_ conn "CREATE TABLE IF NOT EXISTS test_case_failures (id SERIAL PRIMARY KEY, test_case_run_id integer, failure_log text)"
+
+pushToPostgresql :: Options Unwrapped -> (MonoidalMap TestCase Report, Sum Int, Sum Int) -> IO ()
+pushToPostgresql opts (reports, failedRuns, successfulRuns) = do
+  let connInfo =
+        ConnectInfo
+          { connectHost = opts.postgresqlHost,
+            connectPort = opts.postgresqlPort,
+            connectUser = opts.postgresqlUser,
+            connectPassword = opts.postgresqlPassword,
+            connectDatabase = opts.postgresqlDatabase
+          }
+  currentTime <- getCurrentTime
+  bracket (connect connInfo) (close) $ \conn -> do
```

Comment:
Is the linter not complaining about the extra pair of parentheses around `close`?

Reply:
I think we removed that lint warning: https://github.com/wireapp/wire-server/blob/58bcc0c03c99519442ce7979dff12a4bf2cd6d9a/.hlint.yaml#L29-L30
I will fix it though.